### PR TITLE
Add Stargate support

### PR DESCRIPTION
**Pull request by willclarktech**
Opening this as a draft to discuss architecture.

Outstanding issues:
- [x] ~Displaying transaction time in `ExecutionInfo` since `timestamp` isn't available on a Stargate `IndexedTx`.~ #35 
- [x] ~How to get sequence/fee payer/fee granter from a Launchpad transaction in `ui-utils/txs.ts`~ (seems like we don't need these right now)
- [x] React says there are memory leaks. @abefernan You probably know much better than I do how to make this right.
- [x] Give types/type URLs to registry.

### CODE REVIEW

File: src/contexts/ClientContext.tsx [hunk 1]
```diff
@@ -1,9 +1,13 @@
...
+import { CosmWasmClient as LaunchpadClient } from "@cosmjs/cosmwasm";
+import { CosmWasmClient as StargateClient } from "@cosmjs/cosmwasm-stargate";
+import { Registry } from "@cosmjs/proto-signing";
 import React from "react";
 
 export interface ClientContextValue {
   readonly nodeUrl: string;
-  readonly client: CosmWasmClient;
+  readonly launchpadClient: LaunchpadClient;
+  readonly stargateClient?: StargateClient;
```
Comment (webmaster128) [inline src/contexts/ClientContext.tsx:9]: This requires us to always have a `launchpadClient`, which does not make much sense for stargate backends. What about `readonly client: LaunchpadClient | StargateClient;`?

File: src/contexts/ClientContext.tsx [hunk 2]
```diff
@@ -1,9 +1,13 @@
...
+  readonly registry: Registry;
```
Comment (webmaster128) [inline src/contexts/ClientContext.tsx:10]: What about something like this, to add more info in the explorer app context?

```suggestion
  readonly typeRegistry: Registry;
```

File: src/pages/account/AccountPage.tsx [hunk 3]
```diff
@@ -16,33 +17,73 @@ import {
...
                 transactionId: tx.hash,
-                msg: msg,
+                msg: {
+                  fromAddress: msg.value.from_address,
+                  toAddress: msg.value.to_address,
+                  amount: [...msg.value.amount],
+                },
+              });
+            } else {
+              // skip
```
Comment (webmaster128) [inline src/pages/account/AccountPage.tsx:56]: Maybe a `.filter` would be more eadable than this skip code?

File: src/settings/index.ts [hunk 4]
```diff
@@ -9,6 +9,8 @@ export interface DeploymentSettings {
 export interface Settings {
   /** Where do we connect to */
   readonly backend: BackendSettings;
+  /** Are we using Stargate (Launchpad = false) */
+  readonly stargateEnabled: boolean;
```
Comment (webmaster128) [inline src/settings/index.ts:13]: Can you move this to `BackendSettings`? For every given backend, it is either launchpad or stargate

File: src/settings/backend.ts [hunk 5]
```diff
@@ -12,8 +12,10 @@ const heldernetSettings: BackendSettings = {
   nodeUrls: ["https://lcd.heldernet.cosmwasm.com"],
 };
 
+const localPort = process.env.REACT_APP_STARGATE ? "26659" : "1317";
+
 const devnetSettings: BackendSettings = {
-  nodeUrls: ["http://localhost:1317"],
+  nodeUrls: [`http://localhost:${localPort}`],
```
Comment (webmaster128) [inline src/settings/backend.ts:18]: I'd split this in two different backends, once we have the stargate flag in `BackendSettings`

File: src/pages/code/CodePage.tsx [hunk 6]
```diff
@@ -22,36 +21,34 @@ import InstanceRow from "./InstanceRow";
...
-      .searchTx({ tags: makeTags(`message.module=wasm&message.action=store-code&message.code_id=${codeId}`) })
-      .then((results) => {
-        const first = results.find(() => true);
-        setUploadTx(first);
-      })
-      .catch(() => setUploadTx(errorState));
-  }, [clientContext.client, codeId]);
+    ((stargateClient ?? launchpadClient).searchTx({
+      tags: makeTags(`message.module=wasm&message.action=store-code&message.code_id=${codeId}`),
+    }) as Promise<ReadonlyArray<{ readonly hash?: string }>>).then((results) => {
```
Comment (webmaster128) [inline src/pages/code/CodePage.tsx:47]: Why is hash optional here? Isn't it available in all variants of `IndexedTx`?

File: src/pages/account/AccountPage.tsx (hunk 3 above)
Comment (willclarktech) [reply src/pages/account/AccountPage.tsx:56]: This pattern was copied from other parts of the code. I can update.

Comment (webmaster128) [reply src/pages/account/AccountPage.tsx:56]: Both ways are fine for me. Whatever you prefer is good.

File: src/components/FooterRow.tsx [hunk 7]
```diff
@@ -19,25 +19,35 @@ function Separator(): JSX.Element {
...
-  }, [clientContext.client]);
+  const updateHeight = React.useCallback(
+    client === null
+      ? () => {}
+      : () => {
+          client
+            .getHeight()
+            .then(setHeight)
+            .catch(() => setHeight(errorState));
+        },
```
Comment (webmaster128) [inline src/components/FooterRow.tsx:35]: According to https://javascript.info/optional-chaining#short-circuiting, you should be able to do optional chaning with methods as well:

```ts
client?.getHeight() // …
```

File: src/pages/tx/ExecutionInfo.tsx [hunk 8]
```diff
@@ -10,15 +10,15 @@ interface Props {
 }
 
 export function ExecutionInfo({ tx }: Props): JSX.Element {
-  const time = fromRfc3339(tx.timestamp);
+  // const time = fromRfc3339(tx.timestamp);
```
Comment (webmaster128) [inline src/pages/tx/ExecutionInfo.tsx:13]: Is this not available in stargate?

**Review by webmaster128 — APPROVED** [review]

File: src/pages/tx/ExecutionInfo.tsx (hunk 8 above)
Comment (willclarktech) [reply src/pages/tx/ExecutionInfo.tsx:13]: It's not in our `IndexedTx` type: https://github.com/cosmos/cosmjs/blob/master/packages/stargate/src/stargateclient.ts#L34-L42
Maybe it's in the response but we haven't added it so far?

Comment (webmaster128) [reply src/pages/tx/ExecutionInfo.tsx:13]: Would be good to find out. We might need to get the information via the block time if the backend does not provide it for us. It is possible that the REST server did this for us in launchpad times. The time of the block is the only transaction time that exists in the data storage.

Comment (webmaster128) [reply src/pages/tx/ExecutionInfo.tsx:13]: But this should not be blocking the rest of this work to be merged.

Comment (willclarktech) [reply src/pages/tx/ExecutionInfo.tsx:13]: Just checked and this information doesn't seem to be visible in the response.

Comment (webmaster128) [reply src/pages/tx/ExecutionInfo.tsx:13]: Thanks, let's tackle that ndependently: https://github.com/CosmWasm/code-explorer/issues/35

**Pull request merged**